### PR TITLE
Update TS to match Vanilla JS

### PR DESCRIPTION
**Pull request by robpalme**
The TypeScript source code was not equivalent to the JS code due to:

- An extra instantiated function inside `toDrawingColor`
- Use of the legacy TS namespace feature via the legacy `module` keyword
- Various other unnecessary non-type differences

### CODE REVIEW

File: raytracer.ts [hunk 1]
```diff
@@ -35,11 +36,10 @@ class Color {
     static background = Color.black;
     static defaultColor = Color.black;
     static toDrawingColor(c: Color) {
-        var legalize = d => d > 1 ? 1 : d;
         return {
-            r: Math.floor(legalize(c.r) * 255),
-            g: Math.floor(legalize(c.g) * 255),
-            b: Math.floor(legalize(c.b) * 255)
+            r: floor(min(c.r) * 255),
```
Comment (WebReflection) [inline raytracer.ts]: what are these `min` doing?

Comment (robpalme) [reply raytracer.ts]: It was trying to match [the vanilla JS](https://github.com/WebReflection/raytrace/blob/27955b701bc662c87fcefe7287392bc5725d03a5/rt.vanilla.js#L34) but I missed the second argument.  I pushed a fix.

**Comment by WebReflection** [comment]
Current status:

```
raytrace JS: 107.243896484375 ms
rt.vanilla.js:205 raytrace JS: 100.77099609375 ms
rt.vanilla.js:205 raytrace JS: 91.380859375 ms
rt.vanilla.js:205 raytrace JS: 88.196044921875 ms
rt.vanilla.js:205 raytrace JS: 88.62109375 ms
rt.vanilla.js:205 raytrace JS: 89.827880859375 ms

raytrace TS: 123.9990234375 ms
raytracer.ts:284 raytrace TS: 111.6220703125 ms
raytracer.ts:284 raytrace TS: 99.929931640625 ms
raytracer.ts:284 raytrace TS: 99.7998046875 ms
raytracer.ts:284 raytrace TS: 105.597900390625 ms
raytracer.ts:284 raytrace TS: 97.9599609375 ms
```

So the translated TS is somehow still slower, despite you basically copied over my changes but my changes where hand written after the original source code.

Beside the `min(x)` without a `y` which I have commented, the goal is was to show that manual JS is readable, smaller, and faster but of course if we start changing logic or improve implementation or use my JS style in TS too, even if that's not common, I am not sure what we're demonstrating here if not that manually writing JS still produce better results than writing TS with TS style 🤷

Anyway, once your `min` is explained or changed, I'll be hapily merging this into main.

**Comment by robpalme** [comment]
We're demonstrating that the performance difference is due to the optimizations that were applied only to the JS.  This PR shows these same optimizations can equally and easily be applied to TS.  Once you do that, the performance gap disappears.

The goal is to ensure that readers do not get confused into believing that one language has inherently different runtime performance characteristics from the other.  Whilst this repo may not say it explicitly, the social media post that led me here seemed to suggest that JS was inherently faster than TS.

After making just a few alignments in the TS source, the project's `dexnode` runner showed performance parity so I didn't exhaustively align all the code to make it a truly fair test.  Hence why there's still some performance gap on the table when running in a browser.  Further alignment is possible if anyone doubts the languages have performance parity.

**Comment by WebReflection** [comment]
Again: there was a goal with the project I’ve forked and that was ray-tracing.

I wrote a manual JS version for something on the critical path, reducing code branches, simplifying where possible, and the final result was (and still is, even if slightly) faster.

that was the repo idea:

  * types don’t give you better performance, better code does
  * Needing a toolchain to focus less on the task, performance in critical path, and more on type and specific PL features (see the legacy part you removed) can’t really make you more productive
  * Some pattern in TS gets inevitably translated automatically and not necessarily with best performance in mind … your changes demonstrated that and that’s still my point: if you focus on the task you can achieve best performance. If the transpiler or the PL of choice create slow paths, be aware of that and consider JSDoc TS alternative which is never on your way and never changes the meaning of your code.

I hope this clears the original goal and experiment of this repo.

**Comment by WebReflection** [comment]
For history sake, what I've seen here is the following:

  * I apparently wrote better code in vanilla JS so that TS needed to copy over my JS code; it wasn't vice-versa, me copying TS transpiled code over the JS (although the whole writing was of course based on the TS original source code)
  * there were legacy things from TS that TS itself didn't warn me about
  * the first patch had an error (logical, at least, or useless code) in a `min(x)` operation that TS didn't warn not help about
  * there were errors or duplications in the original code that TS didn't warn about neither

As result, JS developers eyes over TS code can produce better results but it's true that is not strictly to blame TS here in particular for anything too slow and the performance are now similar, even if my browsers is on average ~1.1X faster with the hand written JS implementation, in a task where raw performance is crucial too.

**edit** somehow never forget deno core story ... where they abandoned TS to have best performance and more focus on the task ... I see this "story" very similar.

Thanks for the contribution, I will now retire (archive/close) this repository as I don't have time for it and to me it was already done first time I've published it.

**Pull request merged**